### PR TITLE
Use github.com/uber-go/atomic in pkg/util + docs

### PR DESCRIPTION
**Pull request by djmitche**
### What does this PR do?

Adds some documentation about using atomics, and then follows that advice in pkg/util

### Motivation

Safer use of atomics; details in [the dev docs](https://github.com/DataDog/datadog-agent/pull/11716/files?short_path=1d177b1#diff-1d177b1cd865323bfd694812c25f610f9d8e9018f10594d2ca9a39b77acd103f).

### Additional Notes

I'll follow this up with more PRs for other directories.  Eventually we can add a check to ensure sync/atomic isn't used.

### Possible Drawbacks / Trade-offs

None

### Describe how to test/QA your changes

No need.

### Reviewer's Checklist

- [ ] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.
- [ ] Use the `major_change` label if your change either has a major impact on the code base, is impacting multiple teams or is changing important well-established internals of the Agent. This label will be use during QA to make sure each team pay extra attention to the changed behavior. For any customer facing change use a releasenote.
- [ ] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.
- [ ] Changed code has automated tests for its functionality.
- [ ] Adequate QA/testing plan information is provided if the `qa/skip-qa` label is not applied.
- [ ] At least one `team/..` label has been applied, indicating the team(s) that should QA this change.
- [ ] If applicable, docs team has been notified or [an issue has been opened on the documentation repo](https://github.com/DataDog/documentation/issues/new).
- [ ] If applicable, the `need-change/operator` and `need-change/helm` labels have been applied.
- [ ] If applicable, the [config template](https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config_template.yaml) has been updated.

### CODE REVIEW

File: docs/dev/atomics.md [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+Even here, be wary of race conditions, such as assuming that a background goroutine has executed before the test goroutine makes its assertions.
+
+## Use uber-go/atomic
+
+OK, so you've decided to use atomics.
+Use `github.com/uber-go/atomic`, rather than the built-in `sync/atomic` package.
+
+### How
+
+If your atomic is a global variable, you may declare it using its zero value:
```
Comment (mx-psi) [inline docs/dev/atomics.md:24]: I am wondering if it would be easier to follow if we require every atomic variable to be declared as a pointer. I don't have a strong opinion here (thus my approval) but I am putting it out here just in case someone has.

**Review by mx-psi — APPROVED** [review]

**Review by vboulineau — APPROVED** [review]

File: docs/dev/atomics.md (hunk 1 above)
Comment (mx-psi) [reply docs/dev/atomics.md:24]: I don't really know how package-level initialization works or what the concerns are here. This works: https://go.dev/play/p/vpBoGDamE02 and reading this it seems like it would be safe https://go.dev/ref/spec#Package_initialization but I am not entirely sure

File: docs/dev/atomics.md [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+
+```golang
+type FooTracker struct {
+    maxCount *atomic.Uint64
+}
+```
+
+In this case, initialize the pointer in the constructor (`atomic.NewUint64(..)`) and use the `atomic.Uint64` methods to perform atomic operations on the value.
+
+If the additional pointer allocation poses an undue performance burden, include the value as the first element in the struct and include a comment indicating
```
Comment (maycmlee) [inline docs/dev/atomics.md:40]:
```suggestion
If the additional pointer allocation poses an undue performance burden, include the value as the first element in the struct and include a comment indicating:
```

**Review by maycmlee — APPROVED** [review]
Just a small nit!

File: docs/dev/atomics.md [hunk 3]
```diff
@@ -0,0 +1,58 @@
+# Atomic Access
+
+tl;dr: use `github.com/uber-go/atomic` for all atomic access.  Use pointers (`*atomic.Uint64`, etc.) in structs to ensure proper alignment.
```
Comment (mx-psi) [inline docs/dev/atomics.md:3]:
```suggestion
tl;dr: use `go.uber.org/atomic` for all atomic access.  Use pointers (`*atomic.Uint64`, etc.) in structs to ensure proper alignment.
```

File: docs/dev/atomics.md [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+It's tempting to use an atomic to avoid complaints from the race detector, but this is almost always a mistake -- it is merely hiding your race from the race detector.
+Consider carefully why the implementation is racing, and try to address that behavior directly.
+
+The exception to this is tests, where atomics can be useful for sensing some value that you would like to assert on that is manipulated in another goroutine.
+Even here, be wary of race conditions, such as assuming that a background goroutine has executed before the test goroutine makes its assertions.
+
+## Use uber-go/atomic
+
+OK, so you've decided to use atomics.
+Use `github.com/uber-go/atomic`, rather than the built-in `sync/atomic` package.
```
Comment (mx-psi) [inline docs/dev/atomics.md:20]:
```suggestion
Use `go.uber.org/atomic`, rather than the built-in `sync/atomic` package.
```

File: docs/dev/atomics.md [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+
+### How
+
+Always declare atomic types using a pointer.
+The ensures proper alignment.
```
Comment (mx-psi) [inline docs/dev/atomics.md:25]:
```suggestion
This ensures proper alignment.
```

**Review by remeh — APPROVED** [review]
